### PR TITLE
TransparentAddress

### LINKED ISSUE #268: Transparent keys and addresses
We need types for representing transparent addresses and the associated keys. Related to #264.

### PR DESCRIPTION
**Pull request by dconnolly**
And some stubs for other key types, etc.

### CODE REVIEW

File: zebra-chain/src/addresses.rs
Comment (dconnolly) [inline zebra-chain/src/addresses.rs:95]: @hdevalence do we need this, or should this be Serde's Serialize/Deserialize?

Comment (hdevalence) [reply zebra-chain/src/addresses.rs:95]: I think we need it, since there's a canonical format, not one we can invent for ourselves.

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: Since this enum is needed for chain structures, I think it would probably be better to move the definition from zebra-network into, say, `zebra-chain/src/network.rs`, reëxporting it from the crate root, and then updating zebra-network to use `zebra_chain::Network` everywhere.

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: Unless Bitcoin does something very bizarre (real possibility! does it?) I don't think that endianness comes into play here, since it's reading and writing byte sequences rather than integers.

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: Question about data modeling: should there be an `AddressPayloadHash` at all, or should the `PayToScriptHash`/`PayToPublicKeyHash` variants just keep a `[u8; 20]`?

It seems like the `AddressPayloadHash` type doesn't have meaning on its own, but instead is only ever used as part of a `TransparentAddress`. Is this correct? If so, I wonder whether it would be cleaner to eliminate it and have each address variant store its bytes as bytes, since the logic for handling the two different variants will be different.  There might be some duplication in doing the RIPEMD-of-SHA computation, but maybe this only occurs once for each variant (notes 1,2 below).

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: note 1 referenced above

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: note 2 referenced above

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: This comment is a little unclear to me -- from the spec it looks like that is the specific encoding for testnet addresses?

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [inline zebra-chain/src/addresses.rs]: consistency: we should check whether elsewhere in the code uses `pub_key`/`PubKey` or `pubkey`/`Pubkey` and use one of them

**Comment by hdevalence** [comment]
This looks great so far! What are the plans for what will eventually go into this branch before merging?

File: zebra-chain/src/addresses.rs
Comment (dconnolly) [reply zebra-chain/src/addresses.rs]: Yeah `AddressPayloadHash` only ends up being a type that does RIPEMD160(SHA256(bytes)),  that deduplication was kinda nice but the type isn't used anywhere else I think (just checked the spec, none of the shielded address types case about ripemd etc etc). I think I'll get rid of the type and `impl TransparentAddress` to just move the hashing into a method I can call from both `impl From<_>`s?

File: zebra-chain/src/addresses.rs
Comment (dconnolly) [inline zebra-chain/src/addresses.rs]:
```suggestion
```

File: zebra-chain/src/addresses.rs
Comment (dconnolly) [reply zebra-chain/src/addresses.rs]: This was because I know there is also dev network, but I don't know if they have different version bytes or whatever for their addresses, so maybe I should make a note about 'dev net collisions, beware'?

**Comment by dconnolly** [comment]
> This looks great so far! What are the plans for what will eventually go into this branch before merging?

Once the `AddressPayloadHash` is massaged out and the `pub_key/PubKey/pubkey` check are done, I'm ready to merge this unit.

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [reply zebra-chain/src/addresses.rs]: Yeah, that sounds great!

File: zebra-chain/src/addresses.rs
Comment (hdevalence) [reply zebra-chain/src/addresses.rs]: Ah, yeah, I think that would be good to note, but maybe attach the note to the `_ =>` case so that it's clear that it's about possibly unhandled cases? The first time I read it I thought it meant that all non-mainnet versions were read as testnet, so I was confused because it didn't match (my reading of) the code.

File: zebra-chain/src/addresses.rs
Comment (dconnolly) [reply zebra-chain/src/addresses.rs]: `secp256k1` defines `PublicKey`, `ed25519_zebra` exposes `PublicKeyBytes`.  `JoinSplit` has a `pub_key` field.

File: zebra-chain/src/addresses.rs
Comment (dconnolly) [reply zebra-chain/src/addresses.rs]: Should I bring the `magics` in `zebra-network/src/constants.rs` along too? EDIT: I have to because of the inherent impl.

**Review by hdevalence — APPROVED** [review]
LGTM modulo clippy lints!

File: zebra-chain/src/addresses.rs
Comment (daira) [inline zebra-chain/src/addresses.rs:55]: The result of a hash doesn't have a byte order, because it's already defined to be a byte sequence. Byte order applies only when converting between sequences of integers of different element widths. (SHA-256 is big-endian internally but RIPEMD-160 is little-endian internally, if you only use its byte-oriented interface.)

File: zebra-chain/src/addresses.rs
Comment (daira) [inline zebra-chain/src/addresses.rs:103]: Yeah this is only defined by the zcashd implementation (in chainparams.cpp).

File: zebra-chain/src/addresses.rs
Comment (daira) [inline zebra-chain/src/addresses.rs:139]: Adding named constants for this would avoid the repetition of magic values between encoding and decoding, and allow them to be reviewed in one place.

**Pull request merged**
